### PR TITLE
Switch to gen_statem

### PR DESCRIPTION
**Pull request by whatyouhide**
This is a work-in-progress to move Connection from being a GenServer underneath to being a gen_statem. This allows us to simplify the code and not use `gen` and `proc_lib` APIs.

### CODE REVIEW

File: test/connection_test.exs [hunk 1]
```diff
@@ -365,6 +365,11 @@ defmodule ConnectionTest do
              assert_receive {:EXIT, ^pid, {:shutdown, _}}
              Logger.flush()
            end) =~ ~r"error.*GenServer.*\(stop\) shutdown: #Function.*State: 1"sm
```
Comment (whatyouhide) [inline test/connection_test.exs:367]: @josevalim do we still need to do something like this, or is this related to `proc_lib`/`gen`?

Comment (josevalim) [reply test/connection_test.exs:367]: The error message can change, yes.

Comment (whatyouhide) [reply test/connection_test.exs:367]: Right, but in particular I mean, do we still need to do manual logging here? I thought that by implementing `format_status` to return only the `Connection` module state (rather than the gen_statem state), we'd be fine?

Comment (josevalim) [reply test/connection_test.exs:367]: I would perhaps drop the additional logging because both `format_status` and the logging API changed on OTP since this project was released.

**Comment by whatyouhide** [comment]
@josevalim fantastic. Tests are passing over here 🎉

**Comment by hkrutzer** [comment]
I was just looking into this library and it looks good but unmaintained, and complex because of the use of `proc_lib` and such, so this change would be great.

**Comment by josevalim** [comment]
@hkrutzer the library is not unmaintained.

**Comment by josevalim** [comment]
However, it might go unmaintained based on https://github.com/elixir-ecto/db_connection/pull/275. :D Because it may be that rolling your own with `gen_statem` now makes the most sense.

**Comment by whatyouhide** [comment]
Super duper agreed: this PR is complex and risky. Switching db_connection to gen_statem is easier and makes more sense, so let's avoid touching this 👍
